# Advanced Transformation Core
Susan Shin

In [38]:
import pandas as pd
import numpy as np
import os, json

In [39]:
info = pd.read_csv('Data/superhero_info - superhero_info.csv')
info.head()

Hero|Publisher  \
0          A-Bomb|Marvel Comics   
1  Abe Sapien|Dark Horse Comics   
2            Abin Sur|DC Comics   
3     Abomination|Marvel Comics   
4   Absorbing Man|Marvel Comics   

  Gender  \
0   Male   
1   Male   
2   Male   
3   Male   
4   Male   

                Race  \
0              Human   
1      Icthyo Sapien   
2            Ungaran   
3  Human / Radiation   
4              Human   

  Alignment  \
0      good   
1      good   
2      good   
3       bad   
4       bad   

  Hair color  \
0    No Hair   
1    No Hair   
2    No Hair   
3    No Hair   
4    No Hair   

  Eye color  \
0    yellow   
1      blue   
2      blue   
3     green   
4      blue   

  Skin color  \
0    Unknown   
1       blue   
2        red   
3    Unknown   
4    Unknown   

                                   Measurements  
0  {'Height': '203.0 cm', 'Weight': '441.0 kg'}  
1   {'Height': '191.0 cm', 'Weight': '65.0 kg'}  
2   {'Height': '185.0 cm', 'Weight': '90.0 kg'}  
3  {'Height': '203.0 cm', 'Weight': '441.0 kg'}  
4  {'Height': '193.0 cm', 'Weight': '122.0 kg'}

In [40]:
info[['hero_names','publisher']] = info['Hero|Publisher'].str.split('|', expand=True)
info = info.drop(columns = 'Hero|Publisher')
info.head()

Gender  \
0   Male   
1   Male   
2   Male   
3   Male   
4   Male   

                Race  \
0              Human   
1      Icthyo Sapien   
2            Ungaran   
3  Human / Radiation   
4              Human   

  Alignment  \
0      good   
1      good   
2      good   
3       bad   
4       bad   

  Hair color  \
0    No Hair   
1    No Hair   
2    No Hair   
3    No Hair   
4    No Hair   

  Eye color  \
0    yellow   
1      blue   
2      blue   
3     green   
4      blue   

  Skin color  \
0    Unknown   
1       blue   
2        red   
3    Unknown   
4    Unknown   

                                   Measurements  \
0  {'Height': '203.0 cm', 'Weight': '441.0 kg'}   
1   {'Height': '191.0 cm', 'Weight': '65.0 kg'}   
2   {'Height': '185.0 cm', 'Weight': '90.0 kg'}   
3  {'Height': '203.0 cm', 'Weight': '441.0 kg'}   
4  {'Height': '193.0 cm', 'Weight': '122.0 kg'}   

      hero_names  \
0         A-Bomb   
1     Abe Sapien   
2       Abin Sur   
3    Abomination   
4  Absorbing Man   

           publisher  
0      Marvel Comics  
1  Dark Horse Comics  
2          DC Comics  
3      Marvel Comics  
4      Marvel Comics

In [41]:
ms = info.loc[0,"Measurements"]
print(type(ms))
ms

<class 'str'>


"{'Height': '203.0 cm', 'Weight': '441.0 kg'}"

In [42]:
ms = ms.replace("'",'"')
ms

'{"Height": "203.0 cm", "Weight": "441.0 kg"}'

In [43]:
import json
ms_fixed = json.loads(ms)
print(type(ms_fixed))
ms_fixed

<class 'dict'>


{'Height': '203.0 cm', 'Weight': '441.0 kg'}

In [44]:
## use .str.replace to replace all single quotes
info['Measurements'] = info['Measurements'].str.replace("'",'"')
## Apply the json.loads to the full column
info['Measurements'] = info['Measurements'].apply(json.loads)
info['Measurements'].head()

0    {'Height': '203.0 cm', 'Weight': '441.0 kg'}
1     {'Height': '191.0 cm', 'Weight': '65.0 kg'}
2     {'Height': '185.0 cm', 'Weight': '90.0 kg'}
3    {'Height': '203.0 cm', 'Weight': '441.0 kg'}
4    {'Height': '193.0 cm', 'Weight': '122.0 kg'}
Name: Measurements, dtype: object

In [45]:
height_weight = info['Measurements'].apply(pd.Series)
height_weight

Height  \
0    203.0 cm   
1    191.0 cm   
2    185.0 cm   
3    203.0 cm   
4    193.0 cm   
5    185.0 cm   
6    178.0 cm   
7    191.0 cm   
8    188.0 cm   
9    193.0 cm   
10   180.0 cm   
11   244.0 cm   
12   257.0 cm   
13   188.0 cm   
14   183.0 cm   
15   165.0 cm   
16   163.0 cm   
17   183.0 cm   
18   180.0 cm   
19   211.0 cm   
20   183.0 cm   
21   229.0 cm   
22   213.0 cm   
23   178.0 cm   
24   185.0 cm   
25   175.0 cm   
26   183.0 cm   
27   173.0 cm   
28   193.0 cm   
29   185.0 cm   
30   165.0 cm   
31   163.0 cm   
32   183.0 cm   
33   198.0 cm   
34   178.0 cm   
35   168.0 cm   
36   183.0 cm   
37   180.0 cm   
38   183.0 cm   
39   203.0 cm   
40   183.0 cm   
41   165.0 cm   
42   170.0 cm   
43   165.0 cm   
44   168.0 cm   
45   188.0 cm   
46   178.0 cm   
47   178.0 cm   
48   198.0 cm   
49   175.0 cm   
50   180.0 cm   
51   173.0 cm   
52   201.0 cm   
53   188.0 cm   
54   165.0 cm   
55   180.0 cm   
56   198.0 cm   
57   191.0 cm   
58   191.0 cm   
59   188.0 cm   
60   165.0 cm   
61   170.0 cm   
62   178.0 cm   
63   183.0 cm   
64   185.0 cm   
65   170.0 cm   
66   188.0 cm   
67   183.0 cm   
68   170.0 cm   
69   191.0 cm   
70   185.0 cm   
71   188.0 cm   
72   188.0 cm   
73   168.0 cm   
74   165.0 cm   
75   175.0 cm   
76   178.0 cm   
77   218.0 cm   
78   183.0 cm   
79   165.0 cm   
80   196.0 cm   
81   198.0 cm   
82   170.0 cm   
83   183.0 cm   
84   157.0 cm   
85   183.0 cm   
86   170.0 cm   
87   203.0 cm   
88   175.0 cm   
89   183.0 cm   
90   188.0 cm   
91   193.0 cm   
92   198.0 cm   
93   180.0 cm   
94   193.0 cm   
95   175.0 cm   
96   185.0 cm   
97   173.0 cm   
98   175.0 cm   
99   201.0 cm   
100  175.0 cm   
101  180.0 cm   
102  163.0 cm   
103  170.0 cm   
104  175.0 cm   
105  185.0 cm   
106  183.0 cm   
107  226.0 cm   
108  226.0 cm   
109  183.0 cm   
110  183.0 cm   
111  180.0 cm   
112  168.0 cm   
113  198.0 cm   
114  191.0 cm   
115  165.0 cm   
116  183.0 cm   
117  185.0 cm   
118  267.0 cm   
119  168.0 cm   
120  198.0 cm   
121  122.0 cm   
122  173.0 cm   
123  183.0 cm   
124  188.0 cm   
125  185.0 cm   
126  193.0 cm   
127  193.0 cm   
128  185.0 cm   
129  188.0 cm   
130  193.0 cm   
131  198.0 cm   
132  201.0 cm   
133  201.0 cm   
134  188.0 cm   
135  175.0 cm   
136  188.0 cm   
137  173.0 cm   
138  175.0 cm   
139  244.0 cm   
140  196.0 cm   
141  193.0 cm   
142  168.0 cm   
143  180.0 cm   
144  175.0 cm   
145  185.0 cm   
146  178.0 cm   
147  168.0 cm   
148  193.0 cm   
149  188.0 cm   
150  191.0 cm   
151  183.0 cm   
152  196.0 cm   
153  188.0 cm   
154  175.0 cm   
155  975.0 cm   
156  165.0 cm   
157  193.0 cm   
158  173.0 cm   
159  188.0 cm   
160  180.0 cm   
161  183.0 cm   
162  183.0 cm   
163  157.0 cm   
164  183.0 cm   
165  142.0 cm   
166  211.0 cm   
167  180.0 cm   
168  876.0 cm   
169  185.0 cm   
170  183.0 cm   
171  185.0 cm   
172  188.0 cm   
173   62.5 cm   
174  198.0 cm   
175  168.0 cm   
176  175.0 cm   
177  183.0 cm   
178  198.0 cm   
179  178.0 cm   
180  178.0 cm   
181  188.0 cm   
182  180.0 cm   
183  178.0 cm   
184  183.0 cm   
185  178.0 cm   
186  701.0 cm   
187  188.0 cm   
188  188.0 cm   
189  183.0 cm   
190  170.0 cm   
191  183.0 cm   
192  185.0 cm   
193  191.0 cm   
194  165.0 cm   
195  175.0 cm   
196  185.0 cm   
197  180.0 cm   
198  213.0 cm   
199  259.0 cm   
200  173.0 cm   
201  185.0 cm   
202  196.0 cm   
203  180.0 cm   
204  168.0 cm   
205  244.0 cm   
206  178.0 cm   
207  180.0 cm   
208  170.0 cm   
209  175.0 cm   
210  188.0 cm   
211  183.0 cm   
212  173.0 cm   
213  170.0 cm   
214  183.0 cm   
215  180.0 cm   
216  168.0 cm   
217  180.0 cm   
218  198.0 cm   
219  155.0 cm   
220   71.0 cm   
221  178.0 cm   
222  168.0 cm   
223  168.0 cm   
224  170.0 cm   
225  185.0 cm   
226  183.0 cm   
227  196.0 cm   
228  165.0 cm   
229  165.0 cm   
230  287.0 cm   
231  178.0 cm   
232  191.0 cm   
233  173.0 cm   
234  244.0 c

In [46]:
info = pd.concat((info, height_weight), axis = 1)
info.head(2)

Gender  \
0   Male   
1   Male   

            Race  \
0          Human   
1  Icthyo Sapien   

  Alignment  \
0      good   
1      good   

  Hair color  \
0    No Hair   
1    No Hair   

  Eye color  \
0    yellow   
1      blue   

  Skin color  \
0    Unknown   
1       blue   

                                   Measurements  \
0  {'Height': '203.0 cm', 'Weight': '441.0 kg'}   
1   {'Height': '191.0 cm', 'Weight': '65.0 kg'}   

   hero_names  \
0      A-Bomb   
1  Abe Sapien   

           publisher  \
0      Marvel Comics   
1  Dark Horse Comics   

     Height  \
0  203.0 cm   
1  191.0 cm   

     Weight  
0  441.0 kg  
1   65.0 kg

In [47]:
info = info.drop(columns=['Measurements'])
info.head()

Gender  \
0   Male   
1   Male   
2   Male   
3   Male   
4   Male   

                Race  \
0              Human   
1      Icthyo Sapien   
2            Ungaran   
3  Human / Radiation   
4              Human   

  Alignment  \
0      good   
1      good   
2      good   
3       bad   
4       bad   

  Hair color  \
0    No Hair   
1    No Hair   
2    No Hair   
3    No Hair   
4    No Hair   

  Eye color  \
0    yellow   
1      blue   
2      blue   
3     green   
4      blue   

  Skin color  \
0    Unknown   
1       blue   
2        red   
3    Unknown   
4    Unknown   

      hero_names  \
0         A-Bomb   
1     Abe Sapien   
2       Abin Sur   
3    Abomination   
4  Absorbing Man   

           publisher  \
0      Marvel Comics   
1  Dark Horse Comics   
2          DC Comics   
3      Marvel Comics   
4      Marvel Comics   

     Height  \
0  203.0 cm   
1  191.0 cm   
2  185.0 cm   
3  203.0 cm   
4  193.0 cm   

     Weight  
0  441.0 kg  
1   65.0 kg  
2   90.0 kg  
3  441.0 kg  
4  122.0 kg

In [48]:
## save the 2 new columns into the dataframe
info[['Height (in CM)','cm']] = info['Height'].str.split(' ',expand=True)
info[['Weight (in KG)','kg']] = info['Weight'].str.split(' ',expand=True)
info.head(2)

Gender  \
0   Male   
1   Male   

            Race  \
0          Human   
1  Icthyo Sapien   

  Alignment  \
0      good   
1      good   

  Hair color  \
0    No Hair   
1    No Hair   

  Eye color  \
0    yellow   
1      blue   

  Skin color  \
0    Unknown   
1       blue   

   hero_names  \
0      A-Bomb   
1  Abe Sapien   

           publisher  \
0      Marvel Comics   
1  Dark Horse Comics   

     Height  \
0  203.0 cm   
1  191.0 cm   

     Weight  \
0  441.0 kg   
1   65.0 kg   

  Height (in CM)  \
0          203.0   
1          191.0   

   cm  \
0  cm   
1  cm   

  Weight (in KG)  \
0          441.0   
1           65.0   

   kg  
0  kg  
1  kg

In [49]:
## drop the original column 
info = info.drop(columns=['Height', 'Weight', 'cm', 'kg'])
info.head(2)

Gender  \
0   Male   
1   Male   

            Race  \
0          Human   
1  Icthyo Sapien   

  Alignment  \
0      good   
1      good   

  Hair color  \
0    No Hair   
1    No Hair   

  Eye color  \
0    yellow   
1      blue   

  Skin color  \
0    Unknown   
1       blue   

   hero_names  \
0      A-Bomb   
1  Abe Sapien   

           publisher  \
0      Marvel Comics   
1  Dark Horse Comics   

  Height (in CM)  \
0          203.0   
1          191.0   

  Weight (in KG)  
0          441.0  
1           65.0

In [50]:
info['Height (in CM)'] = info['Height (in CM)'].astype(float)
info['Weight (in KG)'] = info['Weight (in KG)'].astype(float)
info.head(2)

Gender  \
0   Male   
1   Male   

            Race  \
0          Human   
1  Icthyo Sapien   

  Alignment  \
0      good   
1      good   

  Hair color  \
0    No Hair   
1    No Hair   

  Eye color  \
0    yellow   
1      blue   

  Skin color  \
0    Unknown   
1       blue   

   hero_names  \
0      A-Bomb   
1  Abe Sapien   

           publisher  \
0      Marvel Comics   
1  Dark Horse Comics   

   Height (in CM)  \
0           203.0   
1           191.0   

   Weight (in KG)  
0           441.0  
1            65.0

In [51]:
info.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 463 entries, 0 to 462
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Gender          463 non-null    object 
 1   Race            463 non-null    object 
 2   Alignment       463 non-null    object 
 3   Hair color      463 non-null    object 
 4   Eye color       463 non-null    object 
 5   Skin color      463 non-null    object 
 6   hero_names      463 non-null    object 
 7   publisher       463 non-null    object 
 8   Height (in CM)  463 non-null    float64
 9   Weight (in KG)  463 non-null    float64
dtypes: float64(2), object(8)
memory usage: 36.3+ KB


In [52]:
powers = pd.read_csv('Data/superhero_powers - superhero_powers.csv')
powers.head()

hero_names  \
0      3-D Man   
1       A-Bomb   
2   Abe Sapien   
3     Abin Sur   
4  Abomination   

                                                                                                                                                                                                            Powers  
0                                                                                                                                                                       Agility,Super Strength,Stamina,Super Speed  
1                                                                                                                       Accelerated Healing,Durability,Longevity,Super Strength,Stamina,Camouflage,Self-Sustenance  
2  Agility,Accelerated Healing,Cold Resistance,Durability,Underwater breathing,Marksmanship,Weapons Master,Longevity,Intelligence,Super Strength,Telepathy,Stamina,Immortality,Reflexes,Enhanced Sight,Sub-Mariner  
3                                                                                                                                                                                               Lantern Power Ring  
4                                                                                                       Accelerated Healing,Intelligence,Super Strength,Stamina,Super Speed,Invulnerability,Animation,Super Breath

In [53]:
df = pd.merge(powers, info, how='inner', on = 'hero_names')
df.head()

hero_names  \
0         A-Bomb   
1     Abe Sapien   
2       Abin Sur   
3    Abomination   
4  Absorbing Man   

                                                                                                                                                                                                            Powers  \
0                                                                                                                       Accelerated Healing,Durability,Longevity,Super Strength,Stamina,Camouflage,Self-Sustenance   
1  Agility,Accelerated Healing,Cold Resistance,Durability,Underwater breathing,Marksmanship,Weapons Master,Longevity,Intelligence,Super Strength,Telepathy,Stamina,Immortality,Reflexes,Enhanced Sight,Sub-Mariner   
2                                                                                                                                                                                               Lantern Power Ring   
3                                                                                                       Accelerated Healing,Intelligence,Super Strength,Stamina,Super Speed,Invulnerability,Animation,Super Breath   
4                  Cold Resistance,Durability,Energy Absorption,Super Strength,Invulnerability,Elemental Transmogrification,Fire Resistance,Natural Armor,Molecular Manipulation,Heat Resistance,Matter Absorption   

  Gender  \
0   Male   
1   Male   
2   Male   
3   Male   
4   Male   

                Race  \
0              Human   
1      Icthyo Sapien   
2            Ungaran   
3  Human / Radiation   
4              Human   

  Alignment  \
0      good   
1      good   
2      good   
3       bad   
4       bad   

  Hair color  \
0    No Hair   
1    No Hair   
2    No Hair   
3    No Hair   
4    No Hair   

  Eye color  \
0    yellow   
1      blue   
2      blue   
3     green   
4      blue   

  Skin color  \
0    Unknown   
1       blue   
2        red   
3    Unknown   
4    Unknown   

           publisher  \
0      Marvel Comics   
1  Dark Horse Comics   
2          DC Comics   
3      Marvel Comics   
4      Marvel Comics   

   Height (in CM)  \
0           203.0   
1           191.0   
2           185.0   
3           203.0   
4           193.0   

   Weight (in KG)  
0           441.0  
1            65.0  
2            90.0  
3           441.0  
4           122.0

In [54]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 463 entries, 0 to 462
Data columns (total 11 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   hero_names      463 non-null    object 
 1   Powers          463 non-null    object 
 2   Gender          463 non-null    object 
 3   Race            463 non-null    object 
 4   Alignment       463 non-null    object 
 5   Hair color      463 non-null    object 
 6   Eye color       463 non-null    object 
 7   Skin color      463 non-null    object 
 8   publisher       463 non-null    object 
 9   Height (in CM)  463 non-null    float64
 10  Weight (in KG)  463 non-null    float64
dtypes: float64(2), object(9)
memory usage: 43.4+ KB


In [55]:
df["Powers"].head()

0                                                                                                                         Accelerated Healing,Durability,Longevity,Super Strength,Stamina,Camouflage,Self-Sustenance
1    Agility,Accelerated Healing,Cold Resistance,Durability,Underwater breathing,Marksmanship,Weapons Master,Longevity,Intelligence,Super Strength,Telepathy,Stamina,Immortality,Reflexes,Enhanced Sight,Sub-Mariner
2                                                                                                                                                                                                 Lantern Power Ring
3                                                                                                         Accelerated Healing,Intelligence,Super Strength,Stamina,Super Speed,Invulnerability,Animation,Super Breath
4                    Cold Resistance,Durability,Energy Absorption,Super Strength,Invulnerability,Elemental Transmogrification,Fire Resistance,Natura

In [56]:
df_pwrs = df["Powers"].str.split(",")

In [57]:
df_pwrs.head()

0                                                                                                                                  [Accelerated Healing, Durability, Longevity, Super Strength, Stamina, Camouflage, Self-Sustenance]
1    [Agility, Accelerated Healing, Cold Resistance, Durability, Underwater breathing, Marksmanship, Weapons Master, Longevity, Intelligence, Super Strength, Telepathy, Stamina, Immortality, Reflexes, Enhanced Sight, Sub-Mariner]
2                                                                                                                                                                                                                [Lantern Power Ring]
3                                                                                                                 [Accelerated Healing, Intelligence, Super Strength, Stamina, Super Speed, Invulnerability, Animation, Super Breath]
4                         [Cold Resistance, Durability, Energy Absorption, Super

In [58]:
exploded = df_pwrs.explode("Powers")
exploded.head()

0    Accelerated Healing
1             Durability
2              Longevity
3         Super Strength
4                Stamina
Name: Powers, dtype: object

In [59]:
## saving the unique values from the exploded column
cols_to_make = exploded.dropna().unique()
cols_to_make

array(['Accelerated Healing', 'Durability', 'Longevity', 'Super Strength',
       'Stamina', 'Camouflage', 'Self-Sustenance', 'Agility',
       'Cold Resistance', 'Underwater breathing', 'Marksmanship',
       'Weapons Master', 'Intelligence', 'Telepathy', 'Immortality',
       'Reflexes', 'Enhanced Sight', 'Sub-Mariner', 'Lantern Power Ring',
       'Super Speed', 'Invulnerability', 'Animation', 'Super Breath',
       'Energy Absorption', 'Elemental Transmogrification',
       'Fire Resistance', 'Natural Armor', 'Molecular Manipulation',
       'Heat Resistance', 'Matter Absorption', 'Stealth', 'Flight',
       'Power Suit', 'Energy Blasts', 'Energy Beams', 'Power Cosmic',
       'Heat Generation', 'Danger Sense', 'Teleportation', 'Phasing',
       'Force Fields', 'Hypnokinesis', 'Energy Manipulation',
       'Invisibility', 'Enhanced Senses', 'Jump', 'Substance Secretion',
       'Natural Weapons', 'Wallcrawling', 'Vision - Thermal',
       'Power Augmentation', 'Cryokinesis', 'Dupli

In [60]:
for col in cols_to_make:
    df[col] = df["Powers"].str.contains(col)
df.head()

/var/folders/th/8ljj2vyx10v7fsgljkwc6hth0000gn/T/ipykernel_12523/2127590937.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  df[col] = df["Powers"].str.contains(col)
/var/folders/th/8ljj2vyx10v7fsgljkwc6hth0000gn/T/ipykernel_12523/2127590937.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  df[col] = df["Powers"].str.contains(col)
/var/folders/th/8ljj2vyx10v7fsgljkwc6hth0000gn/T/ipykernel_12523/2127590937.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many tim

hero_names  \
0         A-Bomb   
1     Abe Sapien   
2       Abin Sur   
3    Abomination   
4  Absorbing Man   

                                                                                                                                                                                                            Powers  \
0                                                                                                                       Accelerated Healing,Durability,Longevity,Super Strength,Stamina,Camouflage,Self-Sustenance   
1  Agility,Accelerated Healing,Cold Resistance,Durability,Underwater breathing,Marksmanship,Weapons Master,Longevity,Intelligence,Super Strength,Telepathy,Stamina,Immortality,Reflexes,Enhanced Sight,Sub-Mariner   
2                                                                                                                                                                                               Lantern Power Ring   
3                                                                                                       Accelerated Healing,Intelligence,Super Strength,Stamina,Super Speed,Invulnerability,Animation,Super Breath   
4                  Cold Resistance,Durability,Energy Absorption,Super Strength,Invulnerability,Elemental Transmogrification,Fire Resistance,Natural Armor,Molecular Manipulation,Heat Resistance,Matter Absorption   

  Gender  \
0   Male   
1   Male   
2   Male   
3   Male   
4   Male   

                Race  \
0              Human   
1      Icthyo Sapien   
2            Ungaran   
3  Human / Radiation   
4              Human   

  Alignment  \
0      good   
1      good   
2      good   
3       bad   
4       bad   

  Hair color  \
0    No Hair   
1    No Hair   
2    No Hair   
3    No Hair   
4    No Hair   

  Eye color  \
0    yellow   
1      blue   
2      blue   
3     green   
4      blue   

  Skin color  \
0    Unknown   
1       blue   
2        red   
3    Unknown   
4    Unknown   

           publisher  \
0      Marvel Comics   
1  Dark Horse Comics   
2          DC Comics   
3      Marvel Comics   
4      Marvel Comics   

   Height (in CM)  \
0           203.0   
1           191.0   
2           185.0   
3           203.0   
4           193.0   

   Weight (in KG)  \
0           441.0   
1            65.0   
2            90.0   
3           441.0   
4           122.0   

   Accelerated Healing  \
0                 True   
1                 True   
2                False   
3                 True   
4                False   

   Durability  \
0        True   
1        True   
2       False   
3       False   
4        True   

   Longevity  \
0       True   
1       True   
2      False   
3      False   
4      False   

   Super Strength  \
0            True   
1            True   
2           False   
3            True   
4            True   

   Stamina  \
0     True   
1     True   
2    False   
3     True   
4    False   

   Camouflage  \
0        True   
1       False   
2       False   
3       False   
4       False   

   Self-Sustenance  \
0             True   
1            False   
2            False   
3            False   
4            False   

   Agility  \
0    False   
1     True   
2    False   
3    False   
4    False   

   Cold Resistance  \
0            False   
1             True   
2            False   
3            False   
4             True   

   Underwater breathing  \
0                 False   
1                  True   
2                 False   
3                 False   
4                 False   

   Marksmanship  \
0         False   
1          True   
2         False   
3         False   
4         False   

   Weapons Master  \
0           False   
1            True   
2           False   
3           False   
4           False   

   Intelligence  \
0         False   
1          True   
2         False   
3          True   
4         False   

   Telepathy  \
0      False   
1       True   
2      False   


# Questions

## Compare the average weight of super powers who have Super Speed to those who do not.

In [61]:
pd.set_option('display.max_column', None)
pd.set_option('display.max_colwidth', None)
pd.set_option('display.width', -1)
pd.set_option('display.max_rows', None)

In [62]:
df_weight = round(df.groupby("Super Speed")["Height (in CM)"].mean(),1)
df_weight

Super Speed
False    186.4
True     189.4
Name: Height (in CM), dtype: float64

Heros with super speed weigh 189.4 compared to 186.4 of those without super speed.

## What is the average height of heroes for each publisher?

In [63]:
df_group = round(df.groupby("publisher")["Height (in CM)"].mean(),1)
df_group

publisher
DC Comics            181.9
Dark Horse Comics    176.9
George Lucas         159.6
Image Comics         211.0
Marvel Comics        191.5
Shueisha             171.5
Star Trek            181.5
Team Epic TV         180.8
Unknown              178.0
Name: Height (in CM), dtype: float64

Above are the average heights for heros by publishers.